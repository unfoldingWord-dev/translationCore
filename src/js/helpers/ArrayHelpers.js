/**
 * @description - Groups consecutive numbers in an array
 * @param {Array} numbers - array of numbers to be grouped
 * @returns {Array} - grouped array of array of consecutive numbers
 */
export const groupConsecutiveNumbers = (numbers) => (
  numbers.reduce(function(accumulator, currentValue, currentIndex, originalArray) {
    if (currentValue) { // ignore undefined entries
      // if this iteration is consecutive to the last, add it to the previous run
      if (currentValue - originalArray[currentIndex - 1] === 1) {
        accumulator[accumulator.length - 1].push(currentValue);
      } else { // the start of a new run including first element
        // create a new subarray with this as the start
        accumulator.push([currentValue]);
      }
    }
    return accumulator; // return state for next iteration
  }, [])
);
/**
 * @description - Deletes indices from an array safely
 * @param {Array} array - array elements to delete from
 * @returns {Array} - the resulting array after indexes were safely removed
 */
export const deleteIndices = (array, indices) => {
  let _array = JSON.parse(JSON.stringify(array));
  indices.sort( (a,b) => b - a );
  indices.forEach(index => {
    if (index >= 0) _array.splice(index, 1);
  });
  return _array;
};
