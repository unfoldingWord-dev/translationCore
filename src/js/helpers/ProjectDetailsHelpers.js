import fs from 'fs-extra';
import path from 'path-extra';
import ospath from 'ospath';
// actions
import * as AlertModalActions from "../actions/AlertModalActions";
import * as OnlineModeConfirmActions from "../actions/OnlineModeConfirmActions";
import * as ProjectInformationCheckActions from "../actions/ProjectInformationCheckActions";
import * as HomeScreenActions from "../actions/HomeScreenActions";
// helpers
import {getTranslate} from "../selectors";
import * as MissingVersesHelpers from './ProjectValidation/MissingVersesHelpers';
import * as GogsApiHelpers from "./GogsApiHelpers";
import * as manifestHelpers from "./manifestHelpers";
import BooksOfTheBible from "../common/BooksOfTheBible";

const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

/**
 * display prompt that project as been renamed
 * @return {Promise} - Returns a promise
 */
export function showRenamedDialog() {
  return ((dispatch, getState) => {
    const { projectDetailsReducer: { projectSaveLocation }} = getState();
    return new Promise(async (resolve) => {
      const translate = getTranslate(getState());
      const projectName = path.basename(projectSaveLocation);
      dispatch(AlertModalActions.openOptionDialog(
        translate('projects.renamed_project', {project: projectName}),
        () => {
          dispatch(AlertModalActions.closeAlertDialog());
          resolve();
        },
        translate('buttons.ok_button')
      ));
    });
  });
}

/**
 * returns true if project name needs to be updated to match spec
 * @param {Object} manifest
 * @param {String} projectSaveLocation
 * @return {Object} - {Boolean} repoNeedsRenaming, {Boolean} newRepoExists, {String} newProjectName
 */
export function shouldProjectNameBeUpdated(manifest, projectSaveLocation) {
  let repoNeedsRenaming = false;
  let newRepoExists = false;
  let newProjectName = null;
  if (projectSaveLocation) {
    newProjectName = generateNewProjectName(manifest);
    const currentProjectName = path.basename(projectSaveLocation);
    repoNeedsRenaming = currentProjectName !== newProjectName;
    const newProjectPath = path.join(path.dirname(projectSaveLocation), newProjectName);
    newRepoExists = fs.existsSync(newProjectPath);
  }
  return { repoNeedsRenaming, newRepoExists, newProjectName };
}

/**
 * returns true if project name already exists in projects
 * @param {String} newProjectName
 * @return {Boolean}
 */
export function doesProjectAlreadyExist(newProjectName) {
  const newProjectPath = path.join(PROJECTS_PATH, newProjectName);
  return fs.existsSync(newProjectPath);
}

/**
 * show user that DCS rename failed, give options
 * @param {String} projectSaveLocation
 * @param {Boolean} createNew - flag that we were doing a create new repo on DCS vs. a rename of the reo
 * @return {Function}
 */
export function showDcsRenameFailure(projectSaveLocation, createNew) {
  return ((dispatch, getState) => {
    const translate = getTranslate(getState());
    const retryText = translate('buttons.retry');
    const continueText = translate('buttons.continue_button');
    const contactHelpDeskText = translate('buttons.contact_helpdesk');
    const projectName = path.basename(projectSaveLocation);
    dispatch(
      AlertModalActions.openOptionDialog(
        translate('projects.dcs_rename_failed', {project:projectName}),
        () => {
          dispatch(AlertModalActions.closeAlertDialog());
          dispatch(handleDcsOperation(createNew, projectSaveLocation)); // retry operation
        },
        retryText,
        continueText,
        contactHelpDeskText,
        () => {
          dispatch(showFeedbackDialog(createNew ? "_.support_dcs_create_new_failed" : "_.support_dcs_rename_failed", () => {
            dispatch(showDcsRenameFailure(projectSaveLocation, createNew)); // reshow alert dialog
          }));
        }
      ));
  });
}

/**
 * format string with details for help desk
 * @param translateKey
 * @return {Function}
 */
export function getFeedbackDetailsForHelpDesk(translateKey) {
  return (async (dispatch, getState) => {
    const state = getState();
    const translate = getTranslate(state);
    const {userdata} = state.loginReducer;
    const { projectSaveLocation } = state.projectDetailsReducer;
    const projectInfo = await GogsApiHelpers.getProjectInfo(projectSaveLocation, userdata);
    return translate(translateKey, projectInfo);
  });
}

/**
 * display the feedback dialog
 * @param {string} translateKey - key of string to use for help desk
 * @param {function} doneCB - callback when feedback dialog closes
 * @return {Function}
 */
export function showFeedbackDialog(translateKey, doneCB = null) {
  return (async (dispatch) => {
    const message = await dispatch(getFeedbackDetailsForHelpDesk(translateKey));
    dispatch(HomeScreenActions.setErrorFeedbackMessage(message)); // put up feedback dialog
    dispatch(HomeScreenActions.setFeedbackCloseCallback(doneCB));
  });
}

const handleDcsRenamePrompOptionDialog = (result, createNewText, projectSaveLocation, resolve) => {
  return ((dispatch, getState) => {
    const createNew = (result === createNewText);
    dispatch(AlertModalActions.closeAlertDialog());
    const {userdata} = getState().loginReducer;
    GogsApiHelpers.changeGitToPointToNewRepo(projectSaveLocation, userdata).then(async () => {
      await dispatch(handleDcsOperation(createNew, projectSaveLocation));
      resolve();
    }).catch((e) => {
      console.log(e);
      resolve();
    });
  });
};

/**
 * handles the renaming on DCS
 * @return {Promise} - Returns a promise
 */
export function doDcsRenamePrompting() {
  return ((dispatch, getState) => {
    const { projectSaveLocation } = getState().projectDetailsReducer;
    return new Promise((resolve) => {
      const translate = getTranslate(getState());
      const renameText = translate('buttons.rename_repo');
      const createNewText = translate('buttons.create_new_repo');
      const projectName = path.basename(projectSaveLocation);
      dispatch(
        AlertModalActions.openOptionDialog(
          translate('projects.dcs_rename_project', {project:projectName}),
          (result) => dispatch(
            handleDcsRenamePrompOptionDialog (result, createNewText, projectSaveLocation, resolve)),
          renameText,
          createNewText,
          null,
          (result) => dispatch(
            handleDcsRenamePrompOptionDialog (result, createNewText, projectSaveLocation, resolve)),
        )
      );
    });
  });
}

/**
 * perform selected action create new or rename project on DCS to match new name
 * @param {boolean} createNew - if true then create new DCS project with current name
 * @param {string} projectSaveLocation
 * @return {Promise<Function>}
 */
export function handleDcsOperation(createNew, projectSaveLocation) {
  return ((dispatch, getState) => {
    return new Promise((resolve) => {
      dispatch(OnlineModeConfirmActions.confirmOnlineAction(
        async () => { // on confirmed
          const {userdata} = getState().loginReducer;
          const projectName = path.basename(projectSaveLocation);
          doesDcsProjectNameAlreadyExist(projectName, userdata).then(async (repoExists) => {
            if (repoExists) {
              dispatch(handleDcsRenameCollision());
            } else {
              if (createNew) {
                try {
                  await GogsApiHelpers.createNewRepo(projectName, projectSaveLocation, userdata);
                } catch (e) {
                  dispatch(showDcsRenameFailure(projectSaveLocation, createNew));
                  console.warn(e);
                }
              } else { // if rename
                try {
                  await GogsApiHelpers.renameRepo(projectName, projectSaveLocation, userdata);
                } catch (e) {
                  dispatch(showDcsRenameFailure(projectSaveLocation, createNew));
                  console.warn(e);
                }
              }
            }
            resolve();
          }).catch((e) => {
            dispatch(showDcsRenameFailure(projectSaveLocation, createNew));
            console.log("exists failure");
            console.log(e);
            resolve();
          });
        },
        () => {
          console.log("cancelled");
          resolve();
        } // on cancel
      ));
    });
  });
}

/**
 * handles the prompting for overwrite/merge of project
 * @return {Promise} - Returns a promise
 */
export function handleDcsRenameCollision() {
  return ((dispatch, getState) => {
    const { projectSaveLocation } = getState().projectDetailsReducer;
    return new Promise((resolve) => {
      const translate = getTranslate(getState());
      const renameText = translate('buttons.rename_local');
      const continueText = translate('buttons.do_not_rename');
      const contactHelpDeskText = translate('buttons.contact_helpdesk');
      const projectName = path.basename(projectSaveLocation);
      dispatch(
        AlertModalActions.openOptionDialog(
          translate('projects.dcs_rename_conflict', {project:projectName}),
          () => {
            dispatch(AlertModalActions.closeAlertDialog());
            dispatch(ProjectInformationCheckActions.openOnlyProjectDetailsScreen(projectSaveLocation));
            resolve();
          },
          renameText,
          continueText,
          contactHelpDeskText,
          () => {
            dispatch(showFeedbackDialog("_.support_dcs_rename_conflict", () => {
              dispatch(handleDcsRenameCollision()); // reshow alert dialog
            }));
            resolve();
          }
        )
      );
    });
  });
}

/**
 * test to see if project name already exists on repo
 * @param {String} newFilename
 * @param {Object} userdata
 * @return {Promise<any>} - resolve returns boolean that file exists
 */
export function doesDcsProjectNameAlreadyExist(newFilename, userdata) {
  return new Promise((resolve, reject) => {
      GogsApiHelpers.findRepo(userdata, newFilename).then(repo => {
        const repoExists = !!repo;
        resolve(repoExists);
      }).catch((e) => {
        console.log(e);
        reject(e);
      });
  });
}

/**
 * separate book and language
 * @param projectName
 * @return {{bookId: string, languageId: *}}
 */
export function getDetailsFromProjectName(projectName) {
  let bookId = "";
  let bookName = "";
  let languageId = "";
  if (projectName) {
    const parts = projectName.split("_");
    languageId = parts[0];
    // we can have a bunch of old formats (e.g. en_act, aaw_php_text_reg) and new format (en_ult_tit_book)
    for (let i = 1; i < parts.length; i++) { // iteratively try the fields to see if valid book ids
      const possibleBookId = parts[i].toLowerCase();
      bookName = BooksOfTheBible.newTestament[possibleBookId];
      if (bookName) {
        bookId = possibleBookId; // if valid bookName use this book id
        break;
      }
    }
  }
  return { bookId, languageId, bookName};
}
/**
 * generate new project name to match spec
 * @param manifest
 * @return {string}
 */
export const generateNewProjectName = (manifest) => {
  let newFilename = '';
  const lang_id = manifest.target_language && manifest.target_language.id ? manifest.target_language.id : '';
  const resourceId = manifest.resource && manifest.resource.id ? manifest.resource.id : '';
  const projectId = manifest.project && manifest.project.id ? manifest.project.id : '';
  const resourceType = "book"; //TODO blm:  hard coded for now
  if (resourceId) {
    newFilename = `${lang_id}_${resourceId}_${projectId}_${resourceType}`;
  } else {
    newFilename = `${lang_id}_${projectId}_${resourceType}`;
  }
  return newFilename.toLowerCase();
};

/**
 * determine what to display for project label and for hover text.  First if there is no project nickname, the project
 *  name is used, else uses projectNickname for project label.  Next if project label is shorter than maximum
 *  length, then full label is displayed and hover text is empty.  Otherwise truncated project label is displayed and
 *  full project label is shown as hover text.
 * @param isProjectLoaded
 * @param projectName
 * @param translate
 * @param projectNickname
 * @param project_max_length
 * @return {{hoverProjectName: String, displayedProjectLabel: String}}
 */
export function getProjectLabel(isProjectLoaded, projectName, translate, projectNickname, project_max_length) {
  const projectLabel = isProjectLoaded ? projectName : translate('project');
  const hoverProjectName = projectNickname || '';
  let displayedProjectLabel = projectLabel || '';
  if (displayedProjectLabel && (displayedProjectLabel.length > project_max_length)) {
    displayedProjectLabel = displayedProjectLabel.substr(0, project_max_length - 1) + '…'; // truncate with ellipsis
  }
  return {hoverProjectName, displayedProjectLabel};
}

/**
 * Gets a tool's progress
 * @param {String} pathToCheckDataFiles
 */
export function getToolProgress(pathToCheckDataFiles) {
  let progress = 0;
  if (fs.existsSync(pathToCheckDataFiles)) {
    let groupDataFiles = fs.readdirSync(pathToCheckDataFiles).filter(file => { // filter out .DS_Store
      return file !== '.DS_Store' && path.extname(file) === '.json';
    });
    let allGroupDataObjects = {};
    groupDataFiles.map((groupDataFileName) => {
      const groupData = fs.readJsonSync(path.join(pathToCheckDataFiles, groupDataFileName));
      allGroupDataObjects[groupDataFileName.replace('.json', '')] = groupData;
    });
    progress = calculateProgress(allGroupDataObjects);
  }
  return progress;
}

/**
 * @description generates the progress percentage
 * @param {object} groupsData - all of the data to calculate percentage from
 * @return {double} - percentage number returned
 */
function calculateProgress(groupsData) {
  let percent;
  const groupIds = Object.keys(groupsData);
  let totalChecks = 0, completedChecks = 0;
  // Loop through all checks and tally completed and totals
  groupIds.forEach(groupId => {
    const groupData = groupsData[groupId];
    groupData.forEach(check => {
      totalChecks += 1;
      // checks are considered completed if selections
      completedChecks += (check.selections) ? 1 : 0;
    });
  });
  // calculate percentage by dividing total by completed
  percent = Math.round(completedChecks / totalChecks * 100) / 100;
  return percent;
}

export function getWordAlignmentProgress(pathToWordAlignmentData, bookId) {
  const groupsObject = {};
  let checked = 0;
  let totalChecks = 0;
  const expectedVerses = MissingVersesHelpers.getExpectedBookVerses(bookId, 'grc', 'ugnt');
  if (fs.existsSync(pathToWordAlignmentData)) {
    let groupDataFiles = fs.readdirSync(pathToWordAlignmentData).filter(file => { // filter out .DS_Store
      return path.extname(file) === '.json';
    });
    groupDataFiles.forEach((chapterFileName) => {
      groupsObject[path.parse(chapterFileName).name] = fs.readJsonSync(path.join(pathToWordAlignmentData, chapterFileName));
    });
    for (let chapterNumber in groupsObject) {
      for (let verseNumber in groupsObject[chapterNumber]) {
        if (!parseInt(verseNumber)) continue;
        const verseDone = isVerseAligned(groupsObject[chapterNumber][verseNumber]);
        if (verseDone) {
          checked++;
        }
      }
    }
    totalChecks = Object.keys(expectedVerses).reduce((chapterTotal, chapterNumber) => {
      return Object.keys(expectedVerses[chapterNumber]).length + chapterTotal;
    }, 0);
  }
  if (totalChecks) {
    return checked / totalChecks;
  }
  return 0;
}

/**
 * checks that verse is aligned, first makes sure that word bank (containing unaligned words) is empty, then double
 *    checks that there are words in verse.  If both of these conditions are true, then we treat it as an aligned verse
 * @param {Object} verseAlignments
 * @return {boolean} true if aligned
 */
export function isVerseAligned(verseAlignments) {
  let aligned = verseAlignments && !verseAlignments.wordBank.length;
  if (aligned) { // if word bank is empty, need to make sure that the verse wasn't empty (no bottom words)
    const foundWords = verseAlignments.alignments.findIndex(alignment => {
      return alignment.bottomWords && alignment.bottomWords.length;
    });
    if (foundWords < 0) { // if verse empty, not aligned
      aligned = false;
    }
  }
  return aligned;
}

export function getWordAlignmentProgressForGroupIndex(projectSaveLocation, bookId, groupIndex) {
  let checked = 0;
  let totalChecks = 0;
  const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
  if(!fs.existsSync(pathToWordAlignmentData)) {
    return 0;
  }
  const chapterNum = groupIndex.id.split('_')[1];
  let groupDataFileName = fs.readdirSync(pathToWordAlignmentData).find(file => { // filter out .DS_Store
    //This will break if we change the wordAlignment tool naming
    //convention of chapter a like chapter_1.json...
    return path.parse(file).name === chapterNum;
  });
  if (groupDataFileName) {
    const groupIndexObject = fs.readJsonSync(path.join(pathToWordAlignmentData, groupDataFileName));
    for (let verseNumber in groupIndexObject) {
      if (parseInt(verseNumber)) {
        const verseDone = isVerseAligned(groupIndexObject[verseNumber]);
        if (verseDone) {
          checked++;
        }
      }
    }
    const expectedVerses = MissingVersesHelpers.getExpectedBookVerses(bookId, 'grc', 'ugnt');
    totalChecks = Object.keys(expectedVerses[chapterNum]).length;
    if (totalChecks) {
      return checked / totalChecks;
    }
  }
  return 0;
}

export function updateProjectFolderName(newProjectName, projectSaveLocation, oldSelectedProjectFileName) {
  const sourcePath = path.join(projectSaveLocation, oldSelectedProjectFileName);
  const destinationPath = path.join(projectSaveLocation, newProjectName);
  if (fs.existsSync(sourcePath) && !fs.existsSync(destinationPath))
    fs.moveSync(sourcePath, destinationPath);
}

/**
 * the folder name is normally the bookId, but if there is not a valid project.id in manifest we fall back to using
 *      the project name.  This function determines if we used the default
 * @param {String} projectFilename
 * @param {String} projectPath
 * @return {String} project folder name
 */
export function getInitialBibleDataFolderName(projectFilename, projectPath) {
  const initialManifest = manifestHelpers.getProjectManifest(projectPath);
  const expectedProjectId = (initialManifest && initialManifest.project && initialManifest.project.id);
  if (expectedProjectId) { // if we have a project id, verify that this is what was actually used for bible data
    const expectedBibleDataPath = path.join(projectPath, expectedProjectId);
    if (fs.existsSync(expectedBibleDataPath)) {
      if(fs.lstatSync(expectedBibleDataPath).isDirectory()) {
        return expectedProjectId;
      }
    }
  }
  return projectFilename; // defaulted to project filename
}

/**
 * if the project.id has been changed in the manifest, we need to move the bible data to new folder
 * @param {Object} manifest
 * @param {String} initialBibleDataFolderName
 * @param {String} projectPath
 */
export function fixBibleDataFolderName(manifest, initialBibleDataFolderName, projectPath) {
  if (manifest && manifest.project && manifest.project.id && (manifest.project.id !== initialBibleDataFolderName)) { // if project.id has changed
    const initialBibleFolderPath = path.join(projectPath, initialBibleDataFolderName);
    const updatedBibleFolderPath = path.join(projectPath, manifest.project.id);
    if(fs.existsSync(initialBibleFolderPath) && ! fs.existsSync(updatedBibleFolderPath)) {
      fs.moveSync(initialBibleFolderPath, updatedBibleFolderPath);
    }
  }
}

