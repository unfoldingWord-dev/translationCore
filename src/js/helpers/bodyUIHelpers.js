export function getIconColorFromIndex(stepIndex) {
  let colorsArray = [];
  colorsArray.push(stepIndex === 0 ? "black" : stepIndex > 0 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 1 ? "black" : 1 > stepIndex ? "grey" : stepIndex > 1 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 2 ? "black" : 2 > stepIndex ? "grey" : stepIndex > 2 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 3 ? "black" : 3 > stepIndex ? "grey" : stepIndex > 3 ? "var(--accent-color-dark)" : "");
  return colorsArray;
}