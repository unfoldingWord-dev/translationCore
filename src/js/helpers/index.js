import { generateTimestamp } from './TimestampGenerator.js';
import * as loadHelpers from './LoadHelpers.js';

export {
  generateTimestamp,
  loadHelpers,
};
