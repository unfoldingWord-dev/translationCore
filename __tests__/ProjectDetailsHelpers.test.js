/* eslint-env jest */
import path from 'path-extra';
import ospath from 'ospath';
import fs from 'fs-extra';
//helpers
import * as ProjectDetailsHelpers from '../src/js/helpers/ProjectDetailsHelpers';
//projects
const alignmentToolProject = '__tests__/fixtures/project/wordAlignment/normal_project';
const emptyAlignmentToolProject = '__tests__/fixtures/project/wordAlignment/empty_project';
const translationWordsProject = '__tests__/fixtures/project/translationWords/normal_project';
const INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
const RESOURCE_PATH = path.join(ospath.home(), 'translationCore', 'resources');


describe('ProjectDetailsHelpers.getWordAlignmentProgress', () => {
  const totalVerses = 46; // for book

  beforeAll(() => {
    const sourcePath = '__tests__/fixtures/project/';
    const destinationPath = '__tests__/fixtures/project/';
    const copyFiles = ['wordAlignment', 'translationWords'];
    fs.__loadFilesIntoMockFs(copyFiles, sourcePath, destinationPath);

    const sourceResourcesPath = path.join('__tests__', 'fixtures', 'resources');
    const resourcesPath = RESOURCE_PATH;
    const copyResourceFiles = ['grc/bibles/ugnt'];
    fs.__loadFilesIntoMockFs(copyResourceFiles, sourceResourcesPath, resourcesPath);
  });

  test('should get the progress of a partially aligned project - 4 verses out of 46', () => {
    const projectSaveLocation = alignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 4;
    const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
    const progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });

  test('should get the progress of a partially aligned project - 3 verses out of 46', () => {
    const projectSaveLocation = alignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 3;
    const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
    unalignChapter1Verse2(pathToWordAlignmentData);
    const progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });

  test('should get the progress of zero for an unaligned project', () => {
    const projectSaveLocation = emptyAlignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 0;
    const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
    const progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });

  test('should get the progress of zero for a new project', () => {
    const projectSaveLocation = emptyAlignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 0;
    const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
    fs.removeSync(pathToWordAlignmentData);
    const progress = ProjectDetailsHelpers.getWordAlignmentProgress(pathToWordAlignmentData, bookId);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });
});

describe('ProjectDetailsHelpers.getToolProgress', () => {
  test('should get the progress for a non alignment tool', () => {
    let toolName = 'translationWords';
    let bookId = 'tit';
    const pathToCheckDataFiles = path.join(translationWordsProject, INDEX_FOLDER_PATH, toolName, bookId);
    expect(ProjectDetailsHelpers.getToolProgress(pathToCheckDataFiles)).toBe(0.06);
  });
});

describe('ProjectDetailsHelpers.getWordAlignmentProgressForGroupIndex', () => {
  const totalVerses = 16; // for chapter 1

  beforeAll(() => {
    const sourcePath = '__tests__/fixtures/project/';
    const destinationPath = '__tests__/fixtures/project/';
    const copyFiles = ['wordAlignment', 'translationWords'];
    fs.__loadFilesIntoMockFs(copyFiles, sourcePath, destinationPath);

    const sourceResourcesPath = path.join('__tests__', 'fixtures', 'resources');
    const resourcesPath = RESOURCE_PATH;
    const copyResourceFiles = ['grc/bibles/ugnt'];
    fs.__loadFilesIntoMockFs(copyResourceFiles, sourceResourcesPath, resourcesPath);
  });

  test('should get the progress of a partially aligned project - 3 verses', () => {
    const projectSaveLocation = alignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 3;
    const groupIndex = { id: 'chapter_1' };
    const progress = ProjectDetailsHelpers.getWordAlignmentProgressForGroupIndex(projectSaveLocation, bookId, groupIndex);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });

  test('should get the progress of a partially aligned project - 2 verses', () => {
    const projectSaveLocation = alignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 2;
    const groupIndex = { id: 'chapter_1' };
    const pathToWordAlignmentData = path.join(projectSaveLocation, '.apps', 'translationCore', 'alignmentData', bookId);
    unalignChapter1Verse2(pathToWordAlignmentData);
    const progress = ProjectDetailsHelpers.getWordAlignmentProgressForGroupIndex(projectSaveLocation, bookId, groupIndex);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });

  test('should get the progress of zero for an unaligned project', () => {
    const projectSaveLocation = emptyAlignmentToolProject;
    const bookId = 'tit';
    const alignedVerses = 0;
    const groupIndex = { id: 'chapter_1' };
    const progress = ProjectDetailsHelpers.getWordAlignmentProgressForGroupIndex(projectSaveLocation, bookId, groupIndex);
    expect(progress).toEqual(alignedVerses/totalVerses);
  });
});

describe('ProjectValidationActions.updateProjectTargetLanguageBookFolderName', () => {
  const projectSaveLocation = 'a/project/path';
  const oldSelectedProjectFileName = 'WRONG_BOOK_ABBR';
  const bookID = 'tit';
  const sourcePath = path.join(projectSaveLocation, oldSelectedProjectFileName);
  const destinationPath = path.join(projectSaveLocation, bookID);
  beforeEach(() => {
    // reset mock filesystem data
    fs.__resetMockFS();
  });
  it('should change the project target language name to the new given one', () => {
    fs.__setMockFS({
      [sourcePath]: ['tit', 'manifest', 'LICENSE']
    });
    expect(fs.existsSync(sourcePath)).toBeTruthy();
    expect(fs.existsSync(destinationPath)).toBeFalsy();
    ProjectDetailsHelpers.updateProjectTargetLanguageBookFolderName(bookID, projectSaveLocation, oldSelectedProjectFileName);
    expect(fs.existsSync(sourcePath)).toBeFalsy();
    expect(fs.existsSync(destinationPath)).toBeTruthy();
  });

  it('should not change the project target language name to the new given one if it already exists', () => {
    fs.__setMockFS({
      [destinationPath]: ['tit', 'manifest', 'LICENSE']
    });
    expect(fs.existsSync(sourcePath)).toBeFalsy();
    expect(fs.existsSync(destinationPath)).toBeTruthy();
    ProjectDetailsHelpers.updateProjectTargetLanguageBookFolderName(bookID, projectSaveLocation, oldSelectedProjectFileName);
    expect(fs.existsSync(sourcePath)).toBeFalsy();
    expect(fs.existsSync(destinationPath)).toBeTruthy();
  });
});

//
// helpers
//

function unalignChapter1Verse2(pathToWordAlignmentData) {
  const chapter1_path = path.join(pathToWordAlignmentData, "1.json");
  const chapter1 = fs.readJSONSync(chapter1_path);
  chapter1[2] = chapter1[3]; // make 2nd verse unaligned
  fs.outputJsonSync(chapter1_path, chapter1);
}

