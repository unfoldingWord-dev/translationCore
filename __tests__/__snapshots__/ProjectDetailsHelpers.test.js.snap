// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetailsHelpers.doDcsRenamePrompting on click create, should call create new 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_project",
    "button1": "buttons.rename_repo",
    "button2": "buttons.create_new_repo",
    "buttonLink": null,
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.doDcsRenamePrompting on click rename, should call rename 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_project",
    "button1": "buttons.rename_repo",
    "button2": "buttons.create_new_repo",
    "buttonLink": null,
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.handleDcsOperation on repo does not exist, should call create new 1`] = `
Array [
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.handleDcsOperation on repo does not exist, should call rename 1`] = `
Array [
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.handleDcsOperation should handle repo exists 1`] = `
Array [
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
  Object {
    "alertText": "projects.dcs_rename_conflict",
    "button1": "buttons.rename_local",
    "button2": "buttons.do_not_rename",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "ProjectInformationCheckActions.openOnlyProjectDetailsScreen",
  },
]
`;

exports[`ProjectDetailsHelpers.handleDcsRenameCollision on click continue, should do nothing 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_conflict",
    "button1": "buttons.rename_local",
    "button2": "buttons.do_not_rename",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "ProjectInformationCheckActions.openOnlyProjectDetailsScreen",
  },
]
`;

exports[`ProjectDetailsHelpers.handleDcsRenameCollision on click rename, should render and open project details 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_conflict",
    "button1": "buttons.rename_local",
    "button2": "buttons.do_not_rename",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "ProjectInformationCheckActions.openOnlyProjectDetailsScreen",
  },
]
`;

exports[`ProjectDetailsHelpers.showDcsRenameFailure on click help desk, should call contactHelpDesk 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_failed",
    "button1": "buttons.retry",
    "button2": "buttons.continue_button",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.showDcsRenameFailure on click retry, should call continue 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_failed",
    "button1": "buttons.retry",
    "button2": "buttons.continue_button",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;

exports[`ProjectDetailsHelpers.showDcsRenameFailure on click retry, should call retry 1`] = `
Array [
  Object {
    "alertText": "projects.dcs_rename_failed",
    "button1": "buttons.retry",
    "button2": "buttons.continue_button",
    "buttonLink": "buttons.contact_helpdesk",
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "CLOSE_ALERT_DIALOG",
  },
  Object {
    "type": "CONFIRM_ONLINE_MODE",
  },
]
`;
